packages/lit-html/src/lib/lit-html.ts

```diff
@@ -12,9 +12,557 @@
+
+    // These regexes represent the five parsing states that we care about and
+    // match the end of the state. Depending on the match, we transition to a
+    // new state. If there's no match, we stay in the same state.
+    // Note that the regexes are stateful. We utilize lastIndex and sync it
+    // across the multiple regexes used. In addition to the five regexes below
+    // we also dynamically create a regex to find the matching end tags for raw
+    // text elements.
+
+    const textRegex = /<((?:!--)|(?:\w*))/g;
+    const commentRegex = /-->/g;
+    const tagRegex = /(>)|(?:(?:[ \x09\x0a\x0c\x0d])([^\0-\x1F\x7F-\x9F "'>=/]+)([ \x09\x0a\x0c\x0d]*=[ \x09\x0a\x0c\x0d]*([^ \x09\x0a\x0c\x0d"'`<>=]|("|')|$)))/g;
+    const singleQuoteAttr = /'/g;
+    const doubleQuoteAttr = /"/g;
+
+    let html = type === 2 ? '<svg>' : '';
+    let quote: string | undefined = undefined;
+    let node: Node | null;
+    let nodeIndex = 0;
+    let bindingIndex = 0;
+
+    // The current parsing state, represented as a reference to one of the
+    // regexes
+    let regex = textRegex;
+
+    for (let i = 0; i < strings.length - 1; i++) {
+      const s = strings[i];
+      // The index of the end of the last attribute. When this is !== -1 at
+      // end of a string, it means we're in q quoted attribute position.
+      let attrNameEnd = -1;
+      let lastIndex = 0;
+
+      // The conditions in this loop handle the current parse state, and the
+      // assignments to the `regex` variable are the state transitions.
+      while (lastIndex < s.length) {
+        // Make sure we start searching from where we previously left off
+        regex.lastIndex = lastIndex;
+        const match = regex.exec(s);
+        if (match === null) {
+          // TODO (justfagnani): add test coverage from spread parts
+          // If we're not in a quoted attribute value, make sure we clear
+          // the attrNameEnd marker
+          if (regex !== singleQuoteAttr && regex !== doubleQuoteAttr) {
+            attrNameEnd = -1;
+          }
+          break;
+        }
+        lastIndex = regex.lastIndex;
+        if (regex === textRegex) {
+          // TODO (justinfagnani): there are more ways to open a comment in
+          // HTML, like `<?` or `</{`. Figure out whether / how to handle that.
+          if (match[1] === '!--') {
+            regex = commentRegex;
+          } else {
+            if (rawTextElement.test(match[1])) {
+              regex = new RegExp(`<\/${match[1]}`);
+            } else {
+              regex = tagRegex;
+            }
+          }
+        } else if (regex === commentRegex) {
+          regex = textRegex;
+        } else if (
+          regex === tagRegex ||
+          regex === doubleQuoteAttr ||
+          regex === singleQuoteAttr
+        ) {
+          if (match[1] === '>') {
+            regex = textRegex;
+          } else if (quote && (match[0] === '"' || match[0] === "'")) {
```

Comment:
I think we should be checking for `=== quote`? Else you could end a double quote with a single quote.
```suggestion
          } else if (match[0] === quote) {
```

Reply:
There are separate regexes for single and double quotes, so we wouldn't get mismatched.

Reply:
Per my second comment, any reason those states are grouped together since their matches appear mutually exclusive?